### PR TITLE
Feature/task 2

### PR DESCRIPTION
**Pull request by VlastimilM**

### CODE REVIEW

File: KenticoListApp/TodoList.Api/Controllers/ListItemsController.cs [hunk 1]
```diff
@@ -1,44 +1,61 @@
...
-
-            if (item.Id != Guid.Empty)
-                return BadRequest("Guid must be empty");
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
 
-            return Created("api/v1/items/?id=300...", await Task.FromResult(SampleItems[0]));
+            return Created("api/v1/items/?id=300...", await Task.FromResult(_repository.Post(item)));
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Api/Controllers/ListItemsController.cs]: The route you return as a location produces an error. It would be much better if it would show me the item I just created using this Post method.
ApiController has an [Url](https://msdn.microsoft.com/en-us/library/system.web.http.apicontroller.url(v=vs.108).aspx?cs-save-lang=1&cs-lang=csharp#code-snippet-1) property which will help you generate URLs. You may also want to inject some class which will create it for you. See "One might..." bullet in the task description.

File: KenticoListApp/TodoList.Api/Controllers/ListItemsController.cs [hunk 2]
```diff
@@ -1,44 +1,61 @@
...
 
         public async Task<IHttpActionResult> PutAsync(ListItem item)
-            => Ok(await Task.FromResult(SampleItems[2]));
+            => Ok(await Task.FromResult(_repository.Put(item)));
+
+        private void ValidatePostItem(ListItem item)
+        {
+            if (item == null)
+            {
+                ModelState.AddModelError(string.Empty, "Item is null");
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Api/Controllers/ListItemsController.cs]: You are using static property Empty on class String, not type string. So, although it works this way, it would look better if you put here String explicitly.

File: KenticoListApp/TodoList.Repositories/ListItemRepository.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    internal class ListItemRepository : IListItemRepository
+    {
+        private static readonly List<ListItem> SampleItems = new List<ListItem>
+        {
+            new ListItem { Id= new Guid("98DBDE18-639E-49A6-8E51-603CEB2AE92D"), Text= "text" },
+            new ListItem { Id = new Guid("1C353E0A-5481-4C31-BD2E-47E1BAF84DBE"), Text = "giraffe" },
+            new ListItem{ Id = new Guid("D69E065C-99B1-4A73-B00C-AD05F071861F"), Text = "updated"}
+        };
+
+        public IEnumerable<ListItem> Get()
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Repositories/ListItemRepository.cs]: Methods in a repository (working with a database) should be async since you do not want to block until database response come.

File: KenticoListApp/TodoList.Repositories/ListItemRepository.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
+using System.Collections.Generic;
+using TodoList.Contracts.Interfaces;
+using TodoList.Contracts.Models;
+
+namespace TodoList.Repositories
+{
+    internal class ListItemRepository : IListItemRepository
+    {
+        private static readonly List<ListItem> SampleItems = new List<ListItem>
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Repositories/ListItemRepository.cs]: List looks more like you want to add or delete items in it. Since you do not intend to add or delete items in SampleItems, it would look better to just create Array for it.

File: KenticoListApp/TodoList.Repositories/ListItemRepository.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+using TodoList.Contracts.Interfaces;
+using TodoList.Contracts.Models;
+
+namespace TodoList.Repositories
+{
+    internal class ListItemRepository : IListItemRepository
+    {
+        private static readonly List<ListItem> SampleItems = new List<ListItem>
+        {
+            new ListItem { Id= new Guid("98DBDE18-639E-49A6-8E51-603CEB2AE92D"), Text= "text" },
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Repositories/ListItemRepository.cs]: So let's find two missing spaces. :)

File: KenticoListApp/TodoList.Repositories/ListItemRepository.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+            new ListItem { Id = new Guid("1C353E0A-5481-4C31-BD2E-47E1BAF84DBE"), Text = "giraffe" },
+            new ListItem{ Id = new Guid("D69E065C-99B1-4A73-B00C-AD05F071861F"), Text = "updated"}
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Repositories/ListItemRepository.cs]: And one missing space here...

File: KenticoListApp/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs [hunk 7]
```diff
@@ -4,163 +4,163 @@
...
-            var actualMessage = error.Message;
 
             Assert.That(responseMessage.StatusCode, Is.EqualTo(HttpStatusCode.BadRequest));
-            Assert.That(actualMessage, Is.EqualTo("Item is null"));
+            Assert.That(error.ModelState.Keys, Is.EqualTo(expectedKeys));
         }
 
         [Test]
         public void PostAsync_WithValidArguments_ReturnsCorrectItemAndStatusCode()
         {
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs]: You should also test returned location. You would find out, that location you are returning now is not correct. :D

File: KenticoListApp/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs [hunk 8]
```diff
@@ -4,163 +4,163 @@
...
             ListItem actualListItem;
             responseMessage.TryGetContentValue(out actualListItem);
 
             Assert.That(responseMessage.StatusCode, Is.EqualTo(HttpStatusCode.OK));
-            Assert.That(actualListItem, Is.EqualTo(expected).Using(_comparer));
+            Assert.That(actualListItem, Is.EqualTo(expectedListItem).UsingListItemComparer());
         }
 
         [Test]
         public void GetAsync_WithoutParams_ReturnsTestItems()
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs]: Forgotten status code check?

File: KenticoListApp/TodoList.Api.Tests/Helpers/EqualConstraintExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using NUnit.Framework.Constraints;
+using TodoList.Contracts.Models;
+
+namespace TodoList.Api.Tests.Helpers
+{
+    public static class EqualConstraintExtensions
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Api.Tests/Helpers/EqualConstraintExtensions.cs]: This can be internal...

File: KenticoListApp/TodoList.Api.Tests/Helpers/ListItemComparer.cs [hunk 10]
```diff
@@ -1,5 +1,5 @@
 ﻿using System.Collections.Generic;
-using TodoList.Api.Models;
+using TodoList.Contracts.Models;
 
 namespace TodoList.Api.Tests.Helpers
 {
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Api.Tests/Helpers/ListItemComparer.cs]: This class can also be internal.

File: KenticoListApp/TodoList.Api/Resolvers/UnityResolver.cs [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+        public IEnumerable<object> GetServices(Type serviceType)
+        {
+            try
+            {
+                return _container.ResolveAll(serviceType);
+            }
+            catch (ResolutionFailedException)
+            {
+                return new List<object>();
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Api/Resolvers/UnityResolver.cs]: Since you are returning just some empty IEnumerable, it would be more readable to return Enumerable.Empty as a sign that it is just some empty Enumerable.

File: KenticoListApp/TodoList.Api/Resolvers/UnityResolver.cs [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+            }
+        }
+
+        public IDependencyScope BeginScope()
+        {
+            var child = _container.CreateChildContainer();
+            return new UnityResolver(child);
+        }
+
+        public void Dispose()
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Api/Resolvers/UnityResolver.cs]: Surely you already know how to fix this.

File: KenticoListApp/TodoList.Api/App_Start/ResponseConfig.cs [hunk 13]
```diff
@@ -0,0 +1,14 @@
+﻿using Newtonsoft.Json.Serialization;
+using System.Web.Http;
+
+namespace TodoList.Api
+{
+    public static class ResponseConfig
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Api/App_Start/ResponseConfig.cs]: Can be internal.

File: KenticoListApp/TodoList.Api/App_Start/RouteConfig.cs [hunk 14]
```diff
@@ -0,0 +1,18 @@
+﻿using System.Web.Http;
+
+namespace TodoList.Api
+{
+    public static class RouteConfig
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Api/App_Start/RouteConfig.cs]: Can be internal.

File: KenticoListApp/TodoList.Api/App_Start/UnityResolverConfig.cs [hunk 15]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Web.Http;
+using Microsoft.Practices.Unity;
+using TodoList.Api.Resolvers;
+using TodoList.Repositories;
+
+namespace TodoList.Api
+{
+    public static class UnityResolverConfig
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Api/App_Start/UnityResolverConfig.cs]: Also can be internal.

File: KenticoListApp/TodoList.Api/App_Start/WebApiConfig.cs [hunk 16]
```diff
@@ -2,13 +2,10 @@
 
 namespace TodoList.Api
 {
-    public static class WebApiConfig
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Api/App_Start/WebApiConfig.cs]: Surely you already know about this.

File: KenticoListApp/TodoList.Api/App_Start/UnityResolverConfig.cs [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+using TodoList.Repositories;
+
+namespace TodoList.Api
+{
+    public static class UnityResolverConfig
+    {
+        public static void Register(HttpConfiguration config)
+        {
+            var container = new UnityContainer()
+                .RegisterRepositoryTypes();
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Api/App_Start/UnityResolverConfig.cs]: Imagine, you want to register classes in another assembly for DI in this architecture. You have to look how other registrations work. Then you find out you have to create another extension on IUnityContainer, which will register your types.
What I would appreciate more (when creating new assemblies for your project) is to create a class implementing some defined interface (e.g. IBootstrapper). That interface would tell me what should I implement. Then use some generic method to tell UnityResolverConfig, that I want to register types also from my class.

File: KenticoListApp/TodoList.Api/Controllers/ListItemsController.cs [hunk 18]
```diff
@@ -1,44 +1,61 @@
...
-            new ListItem(new Guid("D69E065C-99B1-4A73-B00C-AD05F071861F"), "updated")
-        };
+        private readonly IListItemRepository _repository;
 
+        public ListItemsController(IListItemRepository repository)
+        {
+            _repository = repository;
+        }
+        
         public async Task<IHttpActionResult> PostAsync(ListItem item)
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Api/Controllers/ListItemsController.cs]: In my opinion, methods here should be in order Get, Post, Put, Delete. But it is just a matter of taste...

**Review by JiriLojda — CHANGES_REQUESTED** [review]
Good job, there are only a few things, that can be even better. :)

There is only one bigger problem. You have committed many garbage file. You have to double check what you are about to commit. With such files committed into the draft solution, you could break it.
garbagefilesintask2review

Aside from two bigger things and these garbage files, it is all just minor details.

Firstly, a common structure for assemblies is to have sources in Sources folder and tests in Tests folder.

Secondly, please rename your solution to match names of your projects (TodoList). You can also rename project names to match solution name, but it would require much more work. :D

When moving assemblies into folders, you will have to update paths in the solution file. Close Visual Studio and use Notepad++ or you favorite text editor for it, solution file is just a text file with some information about its projects. You will also have to update some paths in project files, those having references on other projects. Use Notepad++ also for this as well.

File: KenticoListApp/TodoList.Contracts/Properties/AssemblyInfo.cs [hunk 19]
```diff
@@ -0,0 +1,39 @@
+﻿using System.Reflection;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+
+[assembly: AssemblyTitle("Contracts")]
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Contracts/Properties/AssemblyInfo.cs]: This should match project name, but it does not, does it?

File: KenticoListApp/TodoList.Contracts/Properties/AssemblyInfo.cs [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("Kentico software")]
+[assembly: AssemblyProduct("Contracts")]
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Contracts/Properties/AssemblyInfo.cs]: Same for this.

File: KenticoListApp/TodoList.Repositories/Properties/AssemblyInfo.cs [hunk 21]
```diff
@@ -0,0 +1,39 @@
+﻿using System.Reflection;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+
+[assembly: AssemblyTitle("ListItemRepository")]
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Repositories/Properties/AssemblyInfo.cs]: And this.

File: KenticoListApp/TodoList.Repositories/Properties/AssemblyInfo.cs [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("Kentico software")]
+[assembly: AssemblyProduct("ListItemRepository")]
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Repositories/Properties/AssemblyInfo.cs]: And this.

**Review by JiriLojda — CHANGES_REQUESTED** [review]

File: KenticoListApp/TodoList.Repositories/ListItemRepository.cs [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+    internal class ListItemRepository : IListItemRepository
+    {
+        private static readonly ListItem[] SampleItems =
+        {
+            new ListItem { Id = new Guid("98DBDE18-639E-49A6-8E51-603CEB2AE92D"), Text = "text" },
+            new ListItem { Id = new Guid("1C353E0A-5481-4C31-BD2E-47E1BAF84DBE"), Text = "giraffe" },
+            new ListItem { Id = new Guid("D69E065C-99B1-4A73-B00C-AD05F071861F"), Text = "updated" }
+        };
+
+        public IEnumerable<ListItem> Get()
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Repositories/ListItemRepository.cs]: There is no need for overload here. Using GetAll would be much more expressive.

File: KenticoListApp/TodoList.Repositories/ListItemRepository.cs [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+        public IEnumerable<ListItem> Get() 
+            => SampleItems;
+
+        public ListItem Get(Guid id) 
+            => SampleItems[0];
+
+        public ListItem Delete(Guid id) 
+            => SampleItems[1];
+
+        public ListItem Put(ListItem item)
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Repositories/ListItemRepository.cs]: Again using Update instead of Put would be much more expressive. You don't have to use name according to  Http verbs here. In fact, you shouldn't use it since this repository can be used for some other Presentation layer.

File: KenticoListApp/TodoList.Repositories/ListItemRepository.cs [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+        public ListItem Get(Guid id) 
+            => SampleItems[0];
+
+        public ListItem Delete(Guid id) 
+            => SampleItems[1];
+
+        public ListItem Put(ListItem item) 
+            => SampleItems[2];
+
+        public ListItem Post(ListItem item)
```
Comment (JiriLojda) [inline KenticoListApp/TodoList.Repositories/ListItemRepository.cs]: Similarly to the previous method, this can be Create instead of Post.

**Review by JiriLojda — CHANGES_REQUESTED** [review]

File: TodoList/Sources/TodoList.Api/Controllers/ListItemsController.cs [hunk 26]
```diff
@@ -18,10 +18,16 @@ public ListItemsController(IListItemRepository repository, IListItemUrlGenerator
         }
 
         public async Task<IHttpActionResult> GetAsync()
-            => Ok(await _repository.GetAllAsync());
```
Comment (PetrSvirak) [inline TodoList/Sources/TodoList.Api/Controllers/ListItemsController.cs]: Why did you do this? I found no mention of something like this in the review nor your commit comment justifying it; also it would strongly suggest against this step. Same applies for the `Put` and `Delete`…

File: TodoList/Sources/TodoList.Api/Controllers/ListItemsController.cs [hunk 27]
```diff
@@ -18,10 +18,16 @@ public ListItemsController(IListItemRepository repository, IListItemUrlGenerator
...
+        {
+            var items = await _repository.GetAllAsync(); 
+            return Ok(items);
+        }
 
         public async Task<IHttpActionResult> GetAsync(Guid id)
-            => Ok(await _repository.GetAsync(id));
+        {
+            var item = await _repository.GetAsync(id);
+            return Ok(item);
```
Comment (PetrSvirak) [inline TodoList/Sources/TodoList.Api/Controllers/ListItemsController.cs]: I do miss `id` validation here. You will want to check for the item of given identifier existence in `task 3`, however, you can surely tell even that now that there is one ID that will (or at least should) never exist in your DB (`Guid.Empty`).

File: TodoList/Sources/TodoList.Api/Services/ListItemUrlGenerator.cs [hunk 28]
```diff
@@ -0,0 +1,20 @@
...
+using TodoList.Contracts.Interfaces;
+using TodoList.Contracts.Models;
+
+namespace TodoList.Api.Services
+{
+    internal class ListItemUrlGenerator : IListItemUrlGenerator
+    {
+        private readonly UrlHelper _helper;
+
+        public ListItemUrlGenerator(HttpRequestMessage message)
```
Comment (PetrSvirak) [inline TodoList/Sources/TodoList.Api/Services/ListItemUrlGenerator.cs:11]: You can even ask for `UrlHelper` helper here, knowing that constructor with most parameters will be used, it will work nicely for you. Just consider this.

File: TodoList/Sources/TodoList.Api/ApiBootstrapper.cs [hunk 29]
```diff
@@ -0,0 +1,17 @@
...
+using TodoList.Contracts.Interfaces;
+
+namespace TodoList.Api
+{
+    internal class ApiBootstrapper : IBootstrapper
+    {
+        public IUnityContainer Register(IUnityContainer container)
+            => container
+                .RegisterType<HttpRequestMessage>(
+                    new InjectionFactory(_ => (HttpRequestMessage) HttpContext.Current.Items["MS_HttpRequestMessage"]))
```
Comment (PetrSvirak) [inline TodoList/Sources/TodoList.Api/ApiBootstrapper.cs]: Move the `HttpRequestMessage` obtainment into a named method. Moreover, you should feel some small urge to justify yourself for such code and put a nice explanatory comment above the named method :stuck_out_tongue:

File: TodoList/Sources/TodoList.Api/ApiBootstrapper.cs [hunk 30]
```diff
@@ -0,0 +1,17 @@
...
+                .RegisterType<IListItemUrlGenerator, ListItemUrlGenerator>();
```
Comment (PetrSvirak) [inline TodoList/Sources/TodoList.Api/ApiBootstrapper.cs]: In both these registrations, scoped lifetime manager is missing.

File: TodoList/Sources/TodoList.Contracts/Interfaces/IListItemUrlGenerator.cs [hunk 31]
```diff
@@ -0,0 +1,9 @@
+﻿using TodoList.Contracts.Models;
+
+namespace TodoList.Contracts.Interfaces
```
Comment (PetrSvirak) [inline TodoList/Sources/TodoList.Contracts/Interfaces/IListItemUrlGenerator.cs]: I'm not completely against this interface being contracts, however, it will always be tightly coupled with .Api, so unless you decided to create a separate assembly for services related to the web (.Services.Web or .Services.Api, ...), I strongly suggest you move this interface right next to its implementation.

File: TodoList/Sources/TodoList.Repositories/RepositoryBootstrapper.cs [hunk 32]
```diff
@@ -0,0 +1,11 @@
+﻿using Microsoft.Practices.Unity;
+using TodoList.Contracts.Interfaces;
+
+namespace TodoList.Repositories
+{
+    public class RepositoryBootstrapper : IBootstrapper
+    {
+        public IUnityContainer Register(IUnityContainer container) 
+            => container.RegisterType<IListItemRepository, ListItemRepository>(new TransientLifetimeManager());
```
Comment (PetrSvirak) [inline TodoList/Sources/TodoList.Repositories/RepositoryBootstrapper.cs]: This is not the correct lifetime manager. You need to use the one that is scoped (based on our verbal agreement and reasoning). Read more on [Unity lifetime manager](https://msdn.microsoft.com/en-us/library/ff660872(v=pandp.20).aspx), please.

File: TodoList/Tests/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs [hunk 33]
```diff
@@ -1,166 +1,178 @@
...
+
+            _controller = new ListItemsController(_repository, _urlGenerator)
+            {
+                Configuration = new HttpConfiguration(),
+                Request = new HttpRequestMessage(),
+            };
+        }
+
+        [Test]
+        public void Delete_ReturnsCorrectResponse()
```
Comment (PetrSvirak) [inline TodoList/Tests/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs]: The tested method is not named `Delete` anymore…

File: TodoList/Tests/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs [hunk 34]
```diff
@@ -1,166 +1,178 @@
...
+
+            var actionResult = _controller.PostAsync(postedListItem).Result;
+            var responseMessage = actionResult.ExecuteAsync(CancellationToken.None).Result;
+            HttpError error;
+            responseMessage.TryGetContentValue(out error);
+            var actualLocation = responseMessage.Headers.Location;
+
+            Assert.That(responseMessage.StatusCode, Is.EqualTo(HttpStatusCode.BadRequest));
+            Assert.That(error.ModelState.Keys, Is.EqualTo(expectedKeys));
+            Assert.That(actualLocation, Is.EqualTo(null));
```
Comment (PetrSvirak) [inline TodoList/Tests/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs]: This actually looks like forget temporary fix. There is much better, more descriptive, way to test for null – simply (assert that <something>) `Is.Null`) – fix all occurences, of course.

File: TodoList/Tests/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs [hunk 35]
```diff
@@ -1,166 +1,178 @@
...
+        }
+
+        [Test]
+        public void PostAsync_WithValidArguments_ReturnsCorrectResponse()
+        {
+            var expectedListItem = new ListItem { Id = _guidOfFirstItem, Text = "text" };
+            var postedListItem = new ListItem { Id = Guid.Empty, Text = "newText" };
+            var expectedLocation = new Uri($"api/v1/ListItems/{_guidOfFirstItem}", UriKind.Relative);
+            _repository.CreateAsync(postedListItem).Returns(expectedListItem);
+            _urlGenerator.GenerateUrl(expectedListItem).Returns($"api/v1/ListItems/{expectedListItem.Id}");
```
Comment (PetrSvirak) [inline TodoList/Tests/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs]: You should actually test whether correct item got passed to the method. Compose the substitute in a way it reads id from the argument passed to the method (you can leverage the [`callInfo` that NSubstitute provides](http://nsubstitute.github.io/help/return-from-function/))

File: TodoList/Tests/TodoList.Api.Tests/Helpers/ListItemComparer.cs [hunk 36]
```diff
@@ -1,18 +1,18 @@
...
-        {
-            return (obj.Id.GetHashCode() * 397) ^ (obj.Text?.GetHashCode() ?? 0);
-        }
-    }
+﻿using System.Collections.Generic;
+using TodoList.Contracts.Models;
+
+namespace TodoList.Api.Tests.Helpers
+{
+    internal class ListItemComparer : IEqualityComparer<ListItem>
```
Comment (PetrSvirak) [inline TodoList/Tests/TodoList.Api.Tests/Helpers/ListItemComparer.cs]: Consider going a step further and -- as this comparer will be used nowhere but in `EqualConstraintExtensions` -- moving it into the `EqualConstraintExtensions.cs` as a private class.

**Review by PetrSvirak — CHANGES_REQUESTED** [review]
It is getting much better quickly. Nice work, guys :-)

#### UnityResolver
There are two unresolved issue with `UnityResolver`
* This way, your `Dispose` is not sufficient. What about inheriting classes? 
* The class has more extensive modifier than it needs, fix this as well, please.

#### Interfaces namespace
The `.Contract.Interfaces` namespace will only clutter you referenced types while it grows (in time).

For example, `IBootstrapper` interfaces will likely be used in other parts (and context) of your application than `IListItemRepository`. 

Moreover, they represent different concepts. The later is implemented in one assembly while the former in all of them. 

In general, it is not bad practice to couple interfaces by assemblies in which they are implemented (or more precisely, couple related things to assemblies and then create separate assemblies for these assemblies' contracts). 

Please, fix this by moving the interfaces into separate, well-named namespaces within the .Contracts assembly.

#### Renegade commit
The d988c02d535dfa4cbed9abb84298da4ee9effe06 has really poor description – it is not cleared what work had been done and why. Omit similar commit(s) in the future, please.

File: TodoList/Sources/TodoList.Api/ApiBootstrapper.cs [hunk 37]
```diff
@@ -2,16 +2,21 @@
...
     {
         public IUnityContainer Register(IUnityContainer container)
             => container
-                .RegisterType<HttpRequestMessage>(
-                    new InjectionFactory(_ => (HttpRequestMessage) HttpContext.Current.Items["MS_HttpRequestMessage"]))
-                .RegisterType<IListItemUrlGenerator, ListItemUrlGenerator>();
+            .RegisterType<HttpRequestMessage>(
+                new HierarchicalLifetimeManager(), 
+                new InjectionFactory(_ => GetCurrentContextRequestMessage()))
+            .RegisterType<IListItemUrlGenerator, ListItemUrlGenerator>(new HierarchicalLifetimeManager());
```
Comment (PetrSvirak) [inline TodoList/Sources/TodoList.Api/ApiBootstrapper.cs]: I kinda liked the previous indentation, however, this one is up to you.

File: TodoList/Sources/TodoList.Api/ApiBootstrapper.cs [hunk 38]
```diff
@@ -2,16 +2,21 @@
 using System.Web;
 using Microsoft.Practices.Unity;
 using TodoList.Api.Services;
-using TodoList.Contracts.Interfaces;
+using TodoList.Contracts.Api;
+using IListItemUrlGenerator = TodoList.Api.Services.IListItemUrlGenerator;
```
Comment (PetrSvirak) [inline TodoList/Sources/TodoList.Api/ApiBootstrapper.cs]: Why do you need this?

File: TodoList/Sources/TodoList.Api/Controllers/ListItemsController.cs [hunk 39]
```diff
@@ -1,78 +1,81 @@
...
+                return BadRequest(ModelState);
+            }
+            return Ok(await _listItemsRepository.GetAsync(id));
+        }
+
+        public async Task<IHttpActionResult> PostAsync(ListItem item)
+        {
+            ValidatePostItem(item);
+
+            if (!ModelState.IsValid)
```
Comment (PetrSvirak) [inline TodoList/Sources/TodoList.Api/Controllers/ListItemsController.cs:40]: You might want to be a bit more consistent between methods. Here, I think the `if` is more related to the `ValidatePostItem` than to item creation, very similar to the case of `ValidateItemId` in `Get`. Thus the free line should rather be present between the `if`'s end and item creation.

File: TodoList/Sources/TodoList.Contracts/Api/IBootstrapper.cs [hunk 40]
```diff
@@ -1,6 +1,6 @@
 ﻿using Microsoft.Practices.Unity;
 
-namespace TodoList.Contracts.Interfaces
+namespace TodoList.Contracts.Api
```
Comment (PetrSvirak) [inline TodoList/Sources/TodoList.Contracts/Api/IBootstrapper.cs]: `Bootstrapper` is in no way connected with (Web) Api while this namespace suggests such meaning.  Move it to a `.Bootstrap` namespace, please. 

The point there is that this interface is used in many (right now two, true) assemblies and they using statements should not express any particular connection to any particular user-facing facility. For example, if there was a WPF application, it would also need to call (all) bootstraps during its startup, yet it would have nothing in common with (web) API…

File: TodoList/Sources/TodoList.Contracts/Interfaces/IListItemUrlGenerator.cs [hunk 41]
```diff
@@ -1,9 +0,0 @@
-﻿using TodoList.Contracts.Models;
```
Comment (PetrSvirak) [inline TodoList/Sources/TodoList.Contracts/Interfaces/IListItemUrlGenerator.cs]: And here you lost your history. @VlastimilM, correct me if I'm wrong, but you it was your commit that doomed the file history, right? So, please, do fix this up so the whole interface lifetime can be seen. None of the commits currently available in the history of the `IListiItemUrl` in the `.Api` (70851b37d9e939e788ba8265b4a4ed847eac2f8e, 0871c8d509db380c0f4375ad087deeda2408eaf3) explains nor justifies its existence.

File: TodoList/Tests/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs [hunk 42]
```diff
@@ -1,178 +1,187 @@
...
+            Assert.That(actualListItem, Is.EqualTo(expectedListItem).UsingListItemComparer());
+        }
+
+        [Test]
+        public void GetAsync_ById_ReturnsCorrectErrorResponse()
+        {
+            var actionResult = _controller.GetAsync(Guid.Empty).Result;
+            var responseMessage = actionResult.ExecuteAsync(CancellationToken.None).Result;
+
+            Assert.That(responseMessage.StatusCode, Is.EqualTo(HttpStatusCode.BadRequest));
```
Comment (PetrSvirak) [inline TodoList/Tests/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs]: The error message (or at least the message's key) should be asserted here as well.

**Review by PetrSvirak — CHANGES_REQUESTED** [review]
There are 3 or 4 nitpicking details that remain. But also 1 incomplete test and 1 task that won't be that easy (chasing lost memories (~ history) is never easy, but I'm sure @VlastimilM will fight it well :crossed_fingers:). Otherwise, the task is complete. We are almost there and you did really nice job overall, thanks.

**Review by PetrSvirak — APPROVED** [review]
